File: financial-tracebility/chaincode-go/chaincode/contract.go
package chaincode

import (
	"fmt"
	"github.com/hyperledger/fabric-contract-api-go/contractapi"
	"github.com/hyperledger/fabric-protos-go/ledger/queryresult"
)

type Contract struct {
	*contractapi.Contract
}

func (c *Contract) Instantiate() {
	fmt.Println("Instantiated")
}

func (c *Contract) Launch(ctx TransactionContextInterface, orderID, price int64, time string) (*Order, error) {
	order, _ := ctx.GetOrderList().GetState(string(orderID))
	if order != nil {
		return nil, fmt.Errorf("existing order with order id %d", orderID)
	}

	order = new(Order)
	order.OrderID = orderID
	order.Price = price
	order.LaunchTime = time
	order.SetState(LAUNCHED)

	err := ctx.GetOrderList().AddState(order)
	if err != nil {
		return nil, err
	}
	return order, nil
}

func (c *Contract) Pay(ctx TransactionContextInterface, orderID int64, time string) (*Order, error) {
	order, err := ctx.GetOrderList().GetState(string(orderID))
	if err != nil {
		return nil, fmt.Errorf("can't get order with id %d", orderID)
	}

	if state := order.GetState(); state != LAUNCHED {
		return nil, fmt.Errorf("order %d's state is %s", orderID, state.String())
	}

	order.PayTime = time
	order.SetState(PAID)

	err = ctx.GetOrderList().UpdateState(order)
	if err != nil {
		return nil, err
	}
	return order, nil
}

func (c *Contract) Dispatch(ctx TransactionContextInterface, orderID int64, time string) (*Order, error) {
	order, err := ctx.GetOrderList().GetState(string(orderID))
	if err != nil {
		return nil, fmt.Errorf("can't get order with id %d", orderID)
	}

	if state := order.GetState(); state != PAID {
		return nil, fmt.Errorf("order %d's state is %s", orderID, state.String())
	}

	order.DispatchTime = time
	order.SetState(DISPATCHING)

	err = ctx.GetOrderList().UpdateState(order)
	if err != nil {
		return nil, err
	}
	return order, nil
}

func (c *Contract) Confirm(ctx TransactionContextInterface, orderID int64, time string) (*Order, error) {
	order, err := ctx.GetOrderList().GetState(string(orderID))
	if err != nil {
		return nil, fmt.Errorf("can't get order with id %d", orderID)
	}

	if state := order.GetState(); state != DISPATCHING {
		return nil, fmt.Errorf("order %d's state is %s", orderID, state.String())
	}

	order.ConfirmTime = time
	order.SetState(CONFIRMED)

	err = ctx.GetOrderList().UpdateState(order)
	if err != nil {
		return nil, err
	}
	return order, nil
}

func (c *Contract) GetOrder(ctx TransactionContextInterface, orderID int64) (*Order, error) {
	order, err := ctx.GetOrderList().GetState(string(orderID))
	if err != nil {
		return nil, fmt.Errorf("can't get order with id %d", orderID)
	}
	return order, nil
}

func (c *Contract) GetHistory(ctx TransactionContextInterface, orderID int64) ([]byte, error) {
	histories, err := ctx.GetStub().GetHistoryForKey(string(orderID))
	if err != nil {
		return nil, fmt.Errorf("can't get history for %d", orderID)
	}
	var result string
	result += "{"
	for histories.HasNext() {
		if history, err := histories.Next(); err != nil {
			break
		} else {
			result += historyToJsonString(history) + ","
		}
	}
	result += "}"
	return []byte(result), nil
}

func historyToJsonString(history *queryresult.KeyModification) string {
	jsonStr := fmt.Sprintf("{\"transactionID\": \"%s\", \"timestamp\": \"%s\", transactionValue: \"%s\"}",
		history.GetTxId(), history.GetTimestamp().String(), history.GetValue())
	return jsonStr
}

File: financial-tracebility/chaincode-java/src/main/java/chaincode/java/Contract.java
package chaincode.java;

import org.hyperledger.fabric.contract.ContractInterface;
import org.hyperledger.fabric.contract.annotation.Transaction;

import chaincode.java.Context;
import org.hyperledger.fabric.shim.ChaincodeStub;
import org.hyperledger.fabric.shim.ledger.KeyModification;

import java.util.List;
import java.util.logging.Logger;

public class Contract implements ContractInterface {
    private final static Logger LOG = Logger.getLogger(Contract.class.getName());

    private final String LAUNCHED = "LAUNCHED";
    private final String PAID = "PAID";
    private final String DISPATCHED = "DISPATCHED";
    private final String CONFIRMED = "CONFIRMED";

    @Override
    public Context createContext(ChaincodeStub stub) {
        return new Context(stub);
    }

    public Contract() {

    }

    @Transaction
    public void instantiate(Context ctx) {
        // No implementation required with this example
        // It could be where data migration is performed, if necessary
        LOG.info("No data migration to perform");
    }

    @Transaction
    public Order launchOrder(Context ctx, String orderID, Integer price, String time) {
        Order order = ctx.getOrderList().getState(orderID);
        if (order != null) {
            throw new RuntimeException("order with id " + orderID + " already exist");
        }

        order = new Order(orderID);
        order.setPrice(price).setlaunchTime(time).setState(LAUNCHED);

        ctx.getOrderList().addState(order);

        return order;
    }

    @Transaction
    public Order payOrder(Context ctx, String orderID, String time) {
        Order order = ctx.getOrderList().getState(orderID);
        if (order == null) {
            throw new RuntimeException("order with id " + orderID + " not launched");
        }

        if (!order.getState().equals(LAUNCHED)) {
            throw new RuntimeException("order with id " + orderID + " 's state is " + order.getState());
        }

        order.setPayTime(time).setState(PAID);
        ctx.getOrderList().updateState(order);

        return order;
    }

    @Transaction
    public Order dispatch(Context ctx, String orderID, String time) {
        Order order = ctx.getOrderList().getState(orderID);

        if (order == null) {
            throw new RuntimeException("order with id " + orderID + " not launched");
        }

        if (!order.getState().equals(PAID)) {
            throw new RuntimeException("order with id " + orderID + " 's state is " + order.getState());
        }

        order.setDispatchTime(time).setState(DISPATCHED);
        ctx.getOrderList().updateState(order);

        return order;
    }

    @Transaction
    public Order confirm(Context ctx, String orderID, String time) {
        Order order = ctx.getOrderList().getState(orderID);
        if (order == null) {
            throw new RuntimeException("order with id " + orderID + " not launched");
        }

        if (!order.getState().equals(DISPATCHED)) {
            throw new RuntimeException("order with id " + orderID + " 's state is " + order.getState());
        }

        order.setConfirmTime(time).setState(CONFIRMED);
        ctx.getOrderList().updateState(order);

        return order;
    }

    @Transaction
    public Order getOrder(Context ctx, String orderID) {
        Order order = ctx.getOrderList().getState(orderID);
        if (order == null) {
            throw new RuntimeException("order with id " + orderID + " not found");
        }
        return order;
    }

    @Transaction
    public String getHistory(Context ctx, String orderID) {
        StringBuilder result = new StringBuilder();
        for (KeyModification km : ctx.getStub().getHistoryForKey(orderID)) {
            result.append("[transaction " + km.getTxId() + "]\n");
            result.append("timestamp: " + km.getTimestamp().toString());
            result.append(km.getStringValue() + "\n");
        }
        return result.toString();
    }
}